src/Arcus.Security.Providers.HashiCorp/HashiCorpSecretProvider.cs

```diff
@@ -53,24 +51,42 @@ public class HashiCorpSecretProvider : ISecretProvider
             Guard.NotNullOrWhitespace(secretPath, nameof(secretPath), "Requires a path where the HashiCorp Vault KeyValue secret engine should look for secrets");
             Guard.For<ArgumentException>(() => !Uri.IsWellFormedUriString(settings.VaultServerUriWithPort, UriKind.RelativeOrAbsolute), "Requires a HashiCorp Vault server URI with HTTP port");
 
-            _options = options;
-            _secretPath = secretPath;
-            _vaultClient = new VaultClient(settings);
-            _logger = logger ?? NullLogger<HashiCorpSecretProvider>.Instance;
+            Options = options;
+            SecretPath = secretPath;
+            VaultClient = new VaultClient(settings);
+            Logger = logger ?? NullLogger<HashiCorpSecretProvider>.Instance;
         }
+        
+        /// <summary>
+        /// Gets the user-configurable options to configure and change the behavior of the HashiCorp KeyValue Vault.
+        /// </summary>
+        protected HashiCorpVaultOptions Options { get; }
+        
+        /// <summary>
+        /// Gets the HashiCorp secret path available in the KeyValue engine where this secret provider should look for secrets.
+        /// </summary>
+        protected string SecretPath { get; }
+        
+        /// <summary>
+        /// Gets the client to interact with the HashiCorp KeyValue Vault, based on the user-provided <see cref="VaultClientSettings"/>.
+        /// </summary>
+        protected IVaultClient VaultClient { get; }
+        
+        /// <summary>
+        /// Gets the logger instance to write diagnostic messages and track HashiCorp Vault dependencies.
+        /// </summary>
+        protected ILogger Logger { get; }
 
         /// <summary>
         /// Retrieves the secret value, based on the given name
         /// </summary>
         /// <param name="secretName">The name of the secret key</param>
         /// <returns>Returns the secret key.</returns>
-        /// <exception cref="ArgumentException">The <paramref name="secretName"/> must not be empty</exception>
-        /// <exception cref="ArgumentNullException">The <paramref name="secretName"/> must not be null</exception>
-        /// <exception cref="SecretNotFoundException">The secret was not found, using the given name</exception>
-        public async Task<string> GetRawSecretAsync(string secretName)
+        /// <exception cref="ArgumentException">Thrown when the <paramref name="secretName"/> is blank.</exception>
+        public virtual async Task<string> GetRawSecretAsync(string secretName)
```

Comment:
I'm still more a fan off dedicated / restricted extension points, but that's maybe just me.
For instance OnBefore / OnAfter overrideable methods.
Or, a default retry-mechanisms, that can be modified by the consumer.   

With making the public methods virtual, you bring off course more power to t he consumer (and more responsabilities).

Reply:
> I'm still more a fan off dedicated / restricted extension points, but that's maybe just me.

I see that 😄 , I think if we look for a broader approach for all Arcus types, it could be quite a job to maintain such methods as people may want to overwrite different things. Especially if you handle return values, alternate inputs... it would also become (I think) quite difficult to know where exactly the overridden call gets executed (I always have that problem when I come across such implementations).
But yeah, that could just be me 😄 .

> With making the public methods virtual, you bring off course more power to t he consumer (and more responsabilities).

True, that's very true. With this, we allow more functionality to be changed, but it's also an approach we can use for every Arcus type, making it more user-friendly in my opinion. We have already received multiple requests for this, that they could override the methods. So, it seems like a logical choice.

But, if you have an alternate example of how this would work, and how we can streamline it across possible other types, feel free to show us!

Reply:
> I'm still more a fan off dedicated / restricted extension points, but that's maybe just me.
> For instance OnBefore / OnAfter overrideable methods.
> Or, a default retry-mechanisms, that can be modified by the consumer.

I can see cases where that's useful indeed, but I mainly see those as "we allow you to do things, but the core is solely for us since it's kind of tricky and we shouldn't mess up".

While in the majority of the cases, you just want to extend what happens - But I'm fine with either, these would just add more methods but that's fine I guess.

Reply:
With regards to https://github.com/arcus-azure/arcus.security/pull/282#issuecomment-878787196, I'd ask @fgheysels instead of me @stijnmoreels

Reply:
Actually, I do acknowledge that there are pro's and con's for both approaches:

- Controlled extension points:  might be enough for 80% of the users, but it is possible that you don't cover the other 20%of the cases.
- Public virtual:  you give tremendous freedom to the user of the library, but it might be easier for the consumer to mess up.  On the other hand: we do not need to take developers by the hand.   Disadvantage:  once you do this, there's no way back.

I don't feel that this is up to me / us to decide.  A discussion with users of the Arcus library or a survey can shed some more light on what the 'Arcus customers' actually want.